src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java

```diff
@@ -0,0 +1,545 @@
+        final TreeNode<IMenu> executionTreeNode = parsingResult.getExecutionTreeNode();
+        if (null == executionTreeNode)
+        {
+            throw new CommandException(getCommandUsage(sender, root));
+        }
+
+        final IMenu executionMenu = executionTreeNode.getData();
+        if (executionMenu.getMenuType().isNavigationMenu())
+        {
+            throw new CommandException(getCommandUsage(sender, executionTreeNode));
+        }
+
+        final ActionMenu actionMenu = (ActionMenu) executionMenu;
+
+        final List<ActionArgument> executionActionArgumentList = parsingResult.getExecutionActionArgumentList();
+        final String badArgument = parsingResult.getBadArgument();
+        throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(executionTreeNode, actionMenu, executionActionArgumentList, badArgument, sender);
+
+        if (sender instanceof EntityPlayer)
+        {
+            final ForgePermissionNodes forgePermissionNode = actionMenu.getForgePermissionNode();
+
+            final EntityPlayer player = (EntityPlayer) sender;
+            if (!PermissionAPI.hasPermission(player.getGameProfile(), forgePermissionNode.getNodeName(), new PlayerContext(player)))
+            {
+                // TODO: Do something if permission check fails.
+                // But we don't have permissions set up yet.
+            }
+        }
+
+        for (final ActionArgument executionActionArgument : executionActionArgumentList)
+        {
+            if (!executionActionArgument.isValueSet())
+            {
+                throw new CommandException(getCommandUsage(sender, executionTreeNode));
+            }
+        }
+
+        final Class<? extends IActionCommand> clazz = actionMenu.getActionCommandClass();
+        try
+        {
+            createInstanceAndExecute(server, sender, actionMenu, clazz);
+        }
+        catch (final InstantiationException | IllegalAccessException e)
+        {
+            final Logger log = LogManager.getLogger();
+            log.error("Unable to instantiate class %s for command %s ", clazz.getName(), actionMenu.getDescription(), e);
+            throw new CommandException("Unable to instantiate class " + clazz.getName() + " for command " + actionMenu.getDescription(), e);
+        }
+    }
+
+    protected void createInstanceAndExecute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final ActionMenu actionMenu,
+            @NotNull final Class<? extends IActionCommand> clazz) throws InstantiationException, IllegalAccessException, CommandException
+    {
+        final IActionCommand actionCommand = clazz.newInstance();
+        actionCommand.execute(server, sender, actionMenu);
+    }
+
+    private static void throwCommandUsageExceptionIfRequiredArgumentsAreNotProvided(@NotNull final TreeNode<IMenu> executionTreeNode, @NotNull final ActionMenu actionMenu,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 34 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)